<!doctype html>
<html lang="zh-CN" data-theme="light">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width,initial-scale=1" />
    <meta name="generator" content="VuePress 2.0.0-rc.26" />
    <meta name="theme" content="VuePress Theme Hope 2.0.0-rc.102" />
    <style>
      :root {
        --vp-c-bg: #fff;
      }

      [data-theme="dark"] {
        --vp-c-bg: #1b1b1f;
      }

      html,
      body {
        background: var(--vp-c-bg);
      }
    </style>
    <script>
      const userMode = localStorage.getItem("vuepress-theme-hope-scheme");
      const systemDarkMode =
        window.matchMedia &&
        window.matchMedia("(prefers-color-scheme: dark)").matches;

      if (userMode === "dark" || (userMode !== "light" && systemDarkMode)) {
        document.documentElement.setAttribute("data-theme", "dark");
      }
    </script>
    <script type="application/ld+json">{"@context":"https://schema.org","@type":"Article","headline":"COMP90014 Algorithms for Bioinformatics","image":[""],"dateModified":"2026-01-07T07:01:42.000Z","author":[{"@type":"Person","name":"Dale","url":"https://github.com/shyu216"}]}</script><meta property="og:url" content="https://shyu216.github.io/knownoevil/knownoevil/study/mscs/24S2/COMP90014.html"><meta property="og:site_name" content="SIHONG's Blog"><meta property="og:title" content="COMP90014 Algorithms for Bioinformatics"><meta property="og:description" content="Prerequisites 1. Introduction to Genomics åŸºå› ç»„å­¦ eukaryotes, prokaryotes, viruses çœŸæ ¸ç”Ÿç‰©ã€åŸæ ¸ç”Ÿç‰©ã€ç—…æ¯’ water 70%, protein 15%, RNA 6%, DNA 1% genome åŸºå› ç»„ bacterium, human, plant, virus ç»†èŒã€..."><meta property="og:type" content="article"><meta property="og:locale" content="zh-CN"><meta property="og:updated_time" content="2026-01-07T07:01:42.000Z"><meta property="article:tag" content="Bioinformatics"><meta property="article:tag" content="Algorithm"><meta property="article:modified_time" content="2026-01-07T07:01:42.000Z"><meta name="application-name" content="SIHONG's Blog"><meta name="mobile-web-app-capable" content="yes"><meta name="theme-color" content="#46bd87"><meta name="apple-mobile-web-app-status-bar-style" content="black"><link rel="icon" href="/knownoevil/favicon.ico"><link rel="manifest" href="/knownoevil/manifest.webmanifest" crossorigin="use-credentials"><link rel="icon" href="/knownoevil/assets/icon/android-chrome-512x512.png" type="image/png" sizes="512x512"><link rel="icon" href="/knownoevil/assets/icon/android-chrome-192x192.png" type="image/png" sizes="192x192"><link rel="apple-touch-icon" href="/knownoevil/assets/icon/apple-touch-icon.png"><title>COMP90014 Algorithms for Bioinformatics | SIHONG's Blog</title><meta name="description" content="Prerequisites 1. Introduction to Genomics åŸºå› ç»„å­¦ eukaryotes, prokaryotes, viruses çœŸæ ¸ç”Ÿç‰©ã€åŸæ ¸ç”Ÿç‰©ã€ç—…æ¯’ water 70%, protein 15%, RNA 6%, DNA 1% genome åŸºå› ç»„ bacterium, human, plant, virus ç»†èŒã€...">
    <link rel="preload" href="/knownoevil/assets/style-Da_-Pejh.css" as="style"><link rel="stylesheet" href="/knownoevil/assets/style-Da_-Pejh.css">
    <link rel="modulepreload" href="/knownoevil/assets/app-Dk-GzZ3g.js"><link rel="modulepreload" href="/knownoevil/assets/COMP90014.html-DDCqR3M6.js"><link rel="modulepreload" href="/knownoevil/assets/plugin-vue_export-helper-DlAUqK2U.js">
    
  </head>
  <body>
    <div id="app"><!--[--><!--[--><!--[--><span tabindex="-1"></span><a href="#main-content" class="vp-skip-link sr-only">è·³è‡³ä¸»è¦å…§å®¹</a><!--]--><!--[--><div class="theme-container external-link-icon has-toc" vp-container><!--[--><header id="navbar" class="vp-navbar" vp-navbar><div class="vp-navbar-start"><button type="button" class="vp-toggle-sidebar-button" title="Toggle Sidebar"><span class="icon"></span></button><!--[--><a class="route-link vp-brand" href="/knownoevil/" aria-label="å¸¦æˆ‘å›å®¶"><img class="vp-nav-logo" src="https://github.com/shyu216.png" alt><!----><span class="vp-site-name hide-in-pad">SIHONG&#39;s Blog</span></a><!--]--></div><div class="vp-navbar-center"><!--[--><nav class="vp-nav-links"><div class="vp-nav-item hide-in-mobile"><a class="route-link route-link-active auto-link" href="/knownoevil/study/" aria-label="Study"><!--[--><i class="vp-icon fas fa-chalkboard-teacher" sizing="height"></i><!--]-->Study<!----></a></div><div class="vp-nav-item hide-in-mobile"><a class="route-link auto-link" href="/knownoevil/coding/" aria-label="Coding"><!--[--><i class="vp-icon fas fa-bug" sizing="height"></i><!--]-->Coding<!----></a></div><div class="vp-nav-item hide-in-mobile"><a class="route-link auto-link" href="/knownoevil/life/" aria-label="Life"><!--[--><i class="vp-icon fas fa-earth-asia" sizing="height"></i><!--]-->Life<!----></a></div></nav><!--]--></div><div class="vp-navbar-end"><!--[--><!----><div class="vp-nav-item vp-action"><a class="vp-action-link" href="https://github.com/shyu216/knownoevil" target="_blank" rel="noopener noreferrer" aria-label="GitHub"><svg xmlns="http://www.w3.org/2000/svg" class="icon github-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="github icon" name="github" style="width:1.25rem;height:1.25rem;vertical-align:middle;"><path d="M511.957 21.333C241.024 21.333 21.333 240.981 21.333 512c0 216.832 140.544 400.725 335.574 465.664 24.49 4.395 32.256-10.07 32.256-23.083 0-11.69.256-44.245 0-85.205-136.448 29.61-164.736-64.64-164.736-64.64-22.315-56.704-54.4-71.765-54.4-71.765-44.587-30.464 3.285-29.824 3.285-29.824 49.195 3.413 75.179 50.517 75.179 50.517 43.776 75.008 114.816 53.333 142.762 40.79 4.523-31.66 17.152-53.377 31.19-65.537-108.971-12.458-223.488-54.485-223.488-242.602 0-53.547 19.114-97.323 50.517-131.67-5.035-12.33-21.93-62.293 4.779-129.834 0 0 41.258-13.184 134.912 50.346a469.803 469.803 0 0 1 122.88-16.554c41.642.213 83.626 5.632 122.88 16.554 93.653-63.488 134.784-50.346 134.784-50.346 26.752 67.541 9.898 117.504 4.864 129.834 31.402 34.347 50.474 78.123 50.474 131.67 0 188.586-114.73 230.016-224.042 242.09 17.578 15.232 33.578 44.672 33.578 90.454v135.85c0 13.142 7.936 27.606 32.854 22.87C862.25 912.597 1002.667 728.747 1002.667 512c0-271.019-219.648-490.667-490.71-490.667z"></path></svg></a></div><div class="vp-nav-item hide-in-mobile"><button type="button" class="vp-color-mode-switch" id="color-mode-switch"><svg xmlns="http://www.w3.org/2000/svg" class="icon auto-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="auto icon" name="auto" style="display:block;"><path d="M512 992C246.92 992 32 777.08 32 512S246.92 32 512 32s480 214.92 480 480-214.92 480-480 480zm0-840c-198.78 0-360 161.22-360 360 0 198.84 161.22 360 360 360s360-161.16 360-360c0-198.78-161.22-360-360-360zm0 660V212c165.72 0 300 134.34 300 300 0 165.72-134.28 300-300 300z"></path></svg><svg xmlns="http://www.w3.org/2000/svg" class="icon dark-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="dark icon" name="dark" style="display:none;"><path d="M524.8 938.667h-4.267a439.893 439.893 0 0 1-313.173-134.4 446.293 446.293 0 0 1-11.093-597.334A432.213 432.213 0 0 1 366.933 90.027a42.667 42.667 0 0 1 45.227 9.386 42.667 42.667 0 0 1 10.24 42.667 358.4 358.4 0 0 0 82.773 375.893 361.387 361.387 0 0 0 376.747 82.774 42.667 42.667 0 0 1 54.187 55.04 433.493 433.493 0 0 1-99.84 154.88 438.613 438.613 0 0 1-311.467 128z"></path></svg><svg xmlns="http://www.w3.org/2000/svg" class="icon light-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="light icon" name="light" style="display:none;"><path d="M952 552h-80a40 40 0 0 1 0-80h80a40 40 0 0 1 0 80zM801.88 280.08a41 41 0 0 1-57.96-57.96l57.96-58a41.04 41.04 0 0 1 58 58l-58 57.96zM512 752a240 240 0 1 1 0-480 240 240 0 0 1 0 480zm0-560a40 40 0 0 1-40-40V72a40 40 0 0 1 80 0v80a40 40 0 0 1-40 40zm-289.88 88.08-58-57.96a41.04 41.04 0 0 1 58-58l57.96 58a41 41 0 0 1-57.96 57.96zM192 512a40 40 0 0 1-40 40H72a40 40 0 0 1 0-80h80a40 40 0 0 1 40 40zm30.12 231.92a41 41 0 0 1 57.96 57.96l-57.96 58a41.04 41.04 0 0 1-58-58l58-57.96zM512 832a40 40 0 0 1 40 40v80a40 40 0 0 1-80 0v-80a40 40 0 0 1 40-40zm289.88-88.08 58 57.96a41.04 41.04 0 0 1-58 58l-57.96-58a41 41 0 0 1 57.96-57.96z"></path></svg></button></div><form class="search-box" role="search"><input type="search" placeholder="æœç´¢" autocomplete="off" spellcheck="false" value><!----></form><!--]--><button type="button" class="vp-toggle-navbar-button" aria-label="Toggle Navbar" aria-expanded="false" aria-controls="nav-screen"><span><span class="vp-top"></span><span class="vp-middle"></span><span class="vp-bottom"></span></span></button></div></header><!----><!--]--><!----><div class="toggle-sidebar-wrapper"><span class="arrow start"></span></div><aside id="sidebar" class="vp-sidebar" vp-sidebar><!----><ul class="vp-sidebar-links"><li><section class="vp-sidebar-group"><button class="vp-sidebar-header clickable" type="button"><i class="vp-icon fas fa-book fa-fw" sizing="both"></i><span class="vp-sidebar-title">CUHK</span><span class="vp-arrow end"></span></button><!----></section></li><li><section class="vp-sidebar-group"><button class="vp-sidebar-header clickable" type="button"><!----><span class="vp-sidebar-title">IELTS</span><span class="vp-arrow end"></span></button><!----></section></li><li><section class="vp-sidebar-group"><button class="vp-sidebar-header clickable active" type="button"><i class="vp-icon fas fa-book fa-fw" sizing="both"></i><span class="vp-sidebar-title">UniMelb</span><span class="vp-arrow down"></span></button><ul class="vp-sidebar-links"><li><section class="vp-sidebar-group"><button class="vp-sidebar-header clickable" type="button"><i class="vp-icon fas fa-book fa-fw" sizing="both"></i><span class="vp-sidebar-title">24S1</span><span class="vp-arrow end"></span></button><!----></section></li><li><section class="vp-sidebar-group"><button class="vp-sidebar-header clickable active" type="button"><i class="vp-icon fas fa-book fa-fw" sizing="both"></i><span class="vp-sidebar-title">24S2</span><span class="vp-arrow down"></span></button><ul class="vp-sidebar-links"><li><a class="route-link route-link-active auto-link vp-sidebar-link active" href="/knownoevil/study/mscs/24S2/COMP90014.html" aria-label="COMP90014"><!--[--><i class="vp-icon fas fa-book-open fa-fw" sizing="both"></i><!--]-->COMP90014<!----></a></li><li><a class="route-link auto-link vp-sidebar-link" href="/knownoevil/study/mscs/24S2/COMP90018.html" aria-label="COMP90018"><!--[--><i class="vp-icon fas fa-book-open fa-fw" sizing="both"></i><!--]-->COMP90018<!----></a></li><li><a class="route-link auto-link vp-sidebar-link" href="/knownoevil/study/mscs/24S2/COMP90044.html" aria-label="COMP90044"><!--[--><i class="vp-icon fas fa-book-open fa-fw" sizing="both"></i><!--]-->COMP90044<!----></a></li><li><a class="route-link auto-link vp-sidebar-link" href="/knownoevil/study/mscs/24S2/COMP90084.html" aria-label="COMP90084"><!--[--><i class="vp-icon fas fa-book-open fa-fw" sizing="both"></i><!--]-->COMP90084<!----></a></li></ul></section></li><li><section class="vp-sidebar-group"><button class="vp-sidebar-header clickable" type="button"><i class="vp-icon fas fa-paint-brush fa-fw" sizing="both"></i><span class="vp-sidebar-title">25 the Research</span><span class="vp-arrow end"></span></button><!----></section></li></ul></section></li></ul><!----></aside><!--[--><main id="main-content" class="vp-page"><!--[--><!----><!----><nav class="vp-breadcrumb disable"></nav><div class="vp-page-title"><h1><i class="vp-icon fas fa-book-open" sizing="height"></i>COMP90014 Algorithms for Bioinformatics</h1><div class="page-info"><span class="page-author-info" aria-label="ä½œè€…ğŸ–Š" data-balloon-pos="up"><svg xmlns="http://www.w3.org/2000/svg" class="icon author-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="author icon" name="author"><path d="M649.6 633.6c86.4-48 147.2-144 147.2-249.6 0-160-128-288-288-288s-288 128-288 288c0 108.8 57.6 201.6 147.2 249.6-121.6 48-214.4 153.6-240 288-3.2 9.6 0 19.2 6.4 25.6 3.2 9.6 12.8 12.8 22.4 12.8h704c9.6 0 19.2-3.2 25.6-12.8 6.4-6.4 9.6-16 6.4-25.6-25.6-134.4-121.6-240-243.2-288z"></path></svg><span><a class="page-author-item" href="https://github.com/shyu216" target="_blank" rel="noopener noreferrer">Dale</a></span><span property="author" content="Dale"></span></span><!----><span class="page-date-info" aria-label="å†™ä½œæ—¥æœŸğŸ“…" data-balloon-pos="up"><svg xmlns="http://www.w3.org/2000/svg" class="icon calendar-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="calendar icon" name="calendar"><path d="M716.4 110.137c0-18.753-14.72-33.473-33.472-33.473-18.753 0-33.473 14.72-33.473 33.473v33.473h66.993v-33.473zm-334.87 0c0-18.753-14.72-33.473-33.473-33.473s-33.52 14.72-33.52 33.473v33.473h66.993v-33.473zm468.81 33.52H716.4v100.465c0 18.753-14.72 33.473-33.472 33.473a33.145 33.145 0 01-33.473-33.473V143.657H381.53v100.465c0 18.753-14.72 33.473-33.473 33.473a33.145 33.145 0 01-33.473-33.473V143.657H180.6A134.314 134.314 0 0046.66 277.595v535.756A134.314 134.314 0 00180.6 947.289h669.74a134.36 134.36 0 00133.94-133.938V277.595a134.314 134.314 0 00-133.94-133.938zm33.473 267.877H147.126a33.145 33.145 0 01-33.473-33.473c0-18.752 14.72-33.473 33.473-33.473h736.687c18.752 0 33.472 14.72 33.472 33.473a33.145 33.145 0 01-33.472 33.473z"></path></svg><span data-allow-mismatch="text">2024/8/1</span><meta property="datePublished" content="2024-08-01T05:48:15.000Z"></span><!----><span class="page-reading-time-info" aria-label="é˜…è¯»æ—¶é—´âŒ›" data-balloon-pos="up"><svg xmlns="http://www.w3.org/2000/svg" class="icon timer-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="timer icon" name="timer"><path d="M799.387 122.15c4.402-2.978 7.38-7.897 7.38-13.463v-1.165c0-8.933-7.38-16.312-16.312-16.312H256.33c-8.933 0-16.311 7.38-16.311 16.312v1.165c0 5.825 2.977 10.874 7.637 13.592 4.143 194.44 97.22 354.963 220.201 392.763-122.204 37.542-214.893 196.511-220.2 389.397-4.661 5.049-7.638 11.651-7.638 19.03v5.825h566.49v-5.825c0-7.379-2.849-13.981-7.509-18.9-5.049-193.016-97.867-351.985-220.2-389.527 123.24-37.67 216.446-198.453 220.588-392.892zM531.16 450.445v352.632c117.674 1.553 211.787 40.778 211.787 88.676H304.097c0-48.286 95.149-87.382 213.728-88.676V450.445c-93.077-3.107-167.901-81.297-167.901-177.093 0-8.803 6.99-15.793 15.793-15.793 8.803 0 15.794 6.99 15.794 15.793 0 80.261 63.69 145.635 142.01 145.635s142.011-65.374 142.011-145.635c0-8.803 6.99-15.793 15.794-15.793s15.793 6.99 15.793 15.793c0 95.019-73.789 172.82-165.96 177.093z"></path></svg><span>å¤§çº¦ 6 åˆ†é’Ÿ</span><meta property="timeRequired" content="PT6M"></span><span class="page-category-info" aria-label="åˆ†ç±»ğŸŒˆ" data-balloon-pos="up"><svg xmlns="http://www.w3.org/2000/svg" class="icon category-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="category icon" name="category"><path d="M148.41 106.992h282.176c22.263 0 40.31 18.048 40.31 40.31V429.48c0 22.263-18.047 40.31-40.31 40.31H148.41c-22.263 0-40.311-18.047-40.311-40.31V147.302c0-22.263 18.048-40.31 40.311-40.31zM147.556 553.478H429.73c22.263 0 40.311 18.048 40.311 40.31v282.176c0 22.263-18.048 40.312-40.31 40.312H147.555c-22.263 0-40.311-18.049-40.311-40.312V593.79c0-22.263 18.048-40.311 40.31-40.311zM593.927 106.992h282.176c22.263 0 40.31 18.048 40.31 40.31V429.48c0 22.263-18.047 40.31-40.31 40.31H593.927c-22.263 0-40.311-18.047-40.311-40.31V147.302c0-22.263 18.048-40.31 40.31-40.31zM730.22 920.502H623.926c-40.925 0-74.22-33.388-74.22-74.425V623.992c0-41.038 33.387-74.424 74.425-74.424h222.085c41.038 0 74.424 33.226 74.424 74.067v114.233c0 10.244-8.304 18.548-18.547 18.548s-18.548-8.304-18.548-18.548V623.635c0-20.388-16.746-36.974-37.33-36.974H624.13c-20.585 0-37.331 16.747-37.331 37.33v222.086c0 20.585 16.654 37.331 37.126 37.331H730.22c10.243 0 18.547 8.304 18.547 18.547 0 10.244-8.304 18.547-18.547 18.547z"></path></svg><!--[--><span class="page-category-item color0 clickable" role="navigation">UniMelb</span><span class="page-category-item color0 clickable" role="navigation">Course</span><span class="page-category-item color0 clickable" role="navigation">24S2</span><!--]--><meta property="articleSection" content="UniMelb,Course,24S2"></span><span class="page-tag-info" aria-label="æ ‡ç­¾ğŸ·" data-balloon-pos="up"><svg xmlns="http://www.w3.org/2000/svg" class="icon tag-icon" viewBox="0 0 1024 1024" fill="currentColor" aria-label="tag icon" name="tag"><path d="M939.902 458.563L910.17 144.567c-1.507-16.272-14.465-29.13-30.737-30.737L565.438 84.098h-.402c-3.215 0-5.726 1.005-7.634 2.913l-470.39 470.39a10.004 10.004 0 000 14.164l365.423 365.424c1.909 1.908 4.42 2.913 7.132 2.913s5.223-1.005 7.132-2.913l470.39-470.39c2.01-2.11 3.014-5.023 2.813-8.036zm-240.067-72.121c-35.458 0-64.286-28.828-64.286-64.286s28.828-64.285 64.286-64.285 64.286 28.828 64.286 64.285-28.829 64.286-64.286 64.286z"></path></svg><!--[--><span class="page-tag-item color0 clickable" role="navigation">Algorithm</span><span class="page-tag-item color0 clickable" role="navigation">Bioinformatics</span><!--]--><meta property="keywords" content="Algorithm,Bioinformatics"></span></div><hr></div><!----><div class="" vp-content><!----><div id="markdown-content"><h2 id="prerequisites-1-introduction-to-genomics-åŸºå› ç»„å­¦" tabindex="-1"><a class="header-anchor" href="#prerequisites-1-introduction-to-genomics-åŸºå› ç»„å­¦"><span>Prerequisites 1. Introduction to Genomics åŸºå› ç»„å­¦</span></a></h2><ul><li>eukaryotes, prokaryotes, viruses <ul><li>çœŸæ ¸ç”Ÿç‰©ã€åŸæ ¸ç”Ÿç‰©ã€ç—…æ¯’</li><li>water 70%, protein 15%, RNA 6%, DNA 1%</li></ul></li><li>genome åŸºå› ç»„ <ul><li>bacterium, human, plant, virus</li><li>ç»†èŒã€äººç±»ã€æ¤ç‰©ã€ç—…æ¯’</li><li>chromosome æŸ“è‰²ä½“</li><li>plasmid è´¨ç²’</li><li>mitochondria çº¿ç²’ä½“</li><li>chloroplast å¶ç»¿ä½“</li></ul></li><li>ploidy å€æ€§ <ul><li>2 copies of the same chromosome, one from each parent</li><li>haploid å•å€ä½“</li><li>diploid äºŒå€ä½“</li><li>polyploid, triploid, tetraploid</li></ul></li><li>molecular biology åˆ†å­ç”Ÿç‰©å­¦ <ul><li>DNA, RNA, protein</li><li>nucleotide æ ¸è‹·é…¸, amino acid æ°¨åŸºé…¸</li><li>gene, genome, transcriptome è½¬å½•ç»„, proteome è›‹ç™½è´¨ç»„</li></ul></li><li>DNA: A-T, C-G <ul><li>double helix åŒèºæ—‹</li><li>sugar-phosphate backbone ç³–ç£·é…¸éª¨æ¶</li><li>5&#39; to 3&#39; direction</li><li>chromosome, gene, exon å¤–æ˜¾å­, intron å†…å«å­, promoter å¯åŠ¨å­, enhancer å¢å¼ºå­, terminator ç»ˆæ­¢å­</li></ul></li><li>protein <ul><li>encoded by gene</li><li>polymer of amino acid monomers æ°¨åŸºé…¸å•ä½“èšåˆç‰©</li><li>cellular functions ç»†èƒåŠŸèƒ½ <ul><li>catalysis å‚¬åŒ–</li><li>structure ç»“æ„</li><li>signalling ä¿¡å·</li></ul></li></ul></li><li>amino acid <ul><li>20 types</li><li>3-letter and 1-letter code</li><li>hydrophobic ç–æ°´, hydrophilic äº²æ°´, charged å¸¦ç”µ</li></ul></li><li>central dogma <ul><li>DNA -&gt; RNA -&gt; protein</li><li>transcription è½¬å½•, translation ç¿»è¯‘</li><li>U (uracil) in RNA instead of T (thymine) in DNA</li><li>ribosome æ ¸ç³–ä½“, tRNA, mRNA</li></ul></li><li>reserve transcription é€†è½¬å½• <ul><li>mRNA -&gt; DNA</li><li>reverse transcriptase é€†è½¬å½•é…¶</li><li>retrovirus é€†è½¬å½•ç—…æ¯’</li></ul></li><li>gene <ul><li>unit of heredity é—ä¼ å•ä½</li><li>2% of human genome codes for genes</li><li>promoter, 5&#39; UTR, coding sequence, 3&#39; UTR, terminator</li><li>coding sequence: exon, intron</li></ul></li><li>variation <ul><li>sequence variation <ul><li>single nucleotide variation (SNV) å•æ ¸è‹·é…¸å˜å¼‚</li><li>single nucleotide polymorphism (SNP) å•æ ¸è‹·é…¸å¤šæ€æ€§</li><li>indels: insertion, deletion</li></ul></li><li>structural variation <ul><li>large-scale (&gt;50bp)</li><li>involving entire sections of DNA</li><li>gain or loss of fucntion (GoF, LoF)</li></ul></li><li>conservation</li></ul></li><li>sequencing technology <ul><li>DNA/RNA extraction -&gt; library preparation -&gt; sequencing -&gt; export fastq files -&gt; QC/analysis</li><li>Illumina, Nanopore</li></ul></li><li>FASTQ files, reads <ul><li>4 lines per read <ul><li>identifier - unique name</li><li>sequence - AGTC</li><li>spacer - +</li><li>quality - confidence score for each DNA base</li></ul></li></ul></li><li>sequencing strategies <ul><li>whole genome sequencing (WGS) <ul><li>all DNA in cells</li></ul></li><li>whole exome sequencing (WES) <ul><li>all exons</li><li>cheap, capture importants</li></ul></li><li>RNA sequencing (RNA-seq) <ul><li>all RNA in cells</li></ul></li><li>scRNA-seq <ul><li>all RNA but idividual cells</li></ul></li><li>Amplicon <ul><li>use PCR target specific region</li></ul></li><li>Pair-end <ul><li>sequence both ends of DNA fragment</li></ul></li></ul></li></ul><h2 id="_1-indexing" tabindex="-1"><a class="header-anchor" href="#_1-indexing"><span>1. Indexing</span></a></h2><ul><li>common classes of algorithms <ul><li>brute force</li><li>divide and conquer</li><li>recursive</li><li>dynamic programming</li><li>greedy</li></ul></li><li>data in bioinformatics <ul><li>FASTA, sequence, genome</li><li>FASTQ, with quality, read</li><li>BAM/SAM, alignment</li><li>Matrix</li><li>VCF, genetic variants</li></ul></li><li>data type <ul><li>primitives: int, float, char</li><li>collections: array, list, set, map</li><li>derived types: class, enum, tuple</li></ul></li><li>data structure <ul><li>arrangement of data in algorithm</li><li>queue, stack, heap, tree, graph</li></ul></li><li>evaluation <ul><li>correctness</li><li>efficiency</li><li>asymptotic complexity æ¸è¿‘å¤æ‚åº¦ (big O notation)</li><li>tractability å¯è§£æ€§ (P, NP, NP-complete)</li></ul></li><li>when complex <ul><li>heuristic</li><li>approximation</li></ul></li><li>indexing <ul><li>pattern matching</li><li>see if something exists</li><li>linear search O(n)</li><li>binary search O(log n)</li><li>hash table O(1)</li></ul></li><li>double hashing (perfect hashing): use another hash function to treat collision</li><li>kmer <ul><li>breaking sequence into smaller pieces</li><li>key of genomic data in indexing</li></ul></li></ul><h2 id="_2-sequence-alignment-and-mapping" tabindex="-1"><a class="header-anchor" href="#_2-sequence-alignment-and-mapping"><span>2. Sequence Alignment and Mapping</span></a></h2><ul><li>why align sequence <ul><li>compare sequence</li><li>assess similarity</li><li>find evolutionary relationships</li><li>identify functionally conserved sequences</li><li>homology åŒæºæ€§ï¼ŒæŒ‡ä¸åŒç”Ÿç‰©ä½“çš„åŸºå› æˆ–è›‹ç™½è´¨åºåˆ—ç”±äºå…±åŒçš„ç¥–å…ˆè€Œå…·æœ‰ç›¸ä¼¼æ€§ï¼Œæœ‰åŠ©äºç†è§£ç”Ÿç‰©è¿›åŒ–ã€åŸºå› åŠŸèƒ½å’Œç»“æ„ <ul><li>homologs åŒæºåŸºå› </li><li>orthologs ç›´ç³» separated by speciation</li><li>paralogs æ—ç³» separated by duplication</li></ul></li><li>phylogeny ç³»ç»Ÿå‘è‚² infer the evolutionary relationships among sequences</li><li>protein function, find regions of known function, infer function</li><li>conservation, find conserved regions åœ¨ä¸åŒç‰©ç§æˆ–ä¸åŒåŸºå› ç»„ä¸­ï¼ŒæŸäº›åŸºå› æˆ–è›‹ç™½è´¨åºåˆ—åœ¨è¿›åŒ–è¿‡ç¨‹ä¸­ä¿æŒä¸å˜æˆ–å˜åŒ–å¾ˆå°çš„åŒºåŸŸï¼Œé€šå¸¸è¿™äº›åºåˆ—åœ¨ç”Ÿç‰©å­¦åŠŸèƒ½ä¸Šå…·æœ‰é‡è¦æ€§</li><li>data searching</li><li>de novo assembly, reconstruct the original sequence</li><li>read mapping, infer variants</li></ul></li><li>how to align <ul><li>pairwise</li><li>BLAST</li></ul></li><li>Hamming distance <ul><li>for equal length sequences, count the number of different characters</li></ul></li><li>kmer distance <ul><li>for unequal length sequences, count the number of different kmers</li></ul></li><li>Levenshtein distance (pairwise) <ul><li>edit distance</li><li>æœ€å°ç¼–è¾‘è·ç¦»ï¼ŒæŒ‡ä¸¤ä¸ªå­—ç¬¦ä¸²ä¹‹é—´ï¼Œç”±ä¸€ä¸ªè½¬æ¢æˆå¦ä¸€ä¸ªæ‰€éœ€çš„æœ€å°‘ç¼–è¾‘æ“ä½œæ¬¡æ•°</li><li>count mismatch, insertion, deletion as transformations</li><li>argmin of <ul><li>(i-1, j-1) + (missmatch 1 or match 0)</li><li>(i-1, j) + (gap 1)</li><li>(i, j-1) + (gap 1)</li></ul></li><li>return the bottom-right cell of the matrix</li></ul></li><li>pairwise distance with penalties <ul><li>apply penalties <ul><li>i.e. match 0, mismatch -1, gap -2</li></ul></li><li>return the bottom-right cell of the matrix</li></ul></li><li>Needleman-Wunsch algorithm (global) <ul><li>return the bottom-right cell of the matrix</li></ul></li><li>Smith-Waterman algorithm (local) <ul><li>return the maximum value in the matrix</li></ul></li><li>semi-global alignment <ul><li>clipped by best offset</li><li>Needleman-Wunsch variant <ul><li>start from max cell, end at reaching top row/left column</li></ul></li></ul></li><li>(scoring) substitution matrix æ›¿æ¢çŸ©é˜µ <ul><li>assign scores to each pair of characters</li><li>why not use fixed scores <ul><li>two types of DNA mutations ä¸¤ç§DNAçªå˜ <ul><li>transition, purine to purine or pyrimidine to pyrimidine, frequently (Aâ†”G, Câ†”T)</li><li>transversion, purine to pyrimidine or vice versa, less likely (Aâ†”C, Aâ†”T, Gâ†”C, Gâ†”T)</li></ul></li></ul></li><li>it reflect <ul><li>chemical similarity</li><li>observed mutation frequencies</li><li>how proteins sequence evolve</li></ul></li></ul></li><li>Point Accepted Mutation (PAM) matrix <ul><li>the number represents the evolutionary distance between the sequences</li><li>higher numbers denote greater distances</li><li>limitation <ul><li>only for closely related sequences and small dataset</li><li>not consider different rates of evolution</li></ul></li></ul></li><li>BLOcks SUbstitution Matrix (BLOSUM) matrix <ul><li>scores derived from frequencies of substitutions in blocks of ungapped local alignments</li><li>BLOSUM62: 62% identity</li></ul></li><li>gap panalty <ul><li>negative score</li><li>fixed / affine</li></ul></li><li>read mapping <ul><li>seed extension <ul><li>combine indexing and alignment</li><li>process <ol><li>index kmer</li><li>find seed</li><li>extend seed using alignment</li><li>return best</li></ol></li></ul></li><li>seed chain align <ul><li>for long, noisy sequences</li><li>process <ol><li>find seeds</li><li>identify colinear chains</li><li>base-level alignments to fill gaps</li><li>return best</li></ol></li></ul></li></ul></li><li>Basic Local Alignment Search Tool (BLAST) <ul><li>efficient heuristic <ul><li>DNA translated to protein seq for use <ul><li>Amino Acid (AA) sequences more conserved than DNA</li><li>degeneracy / codon wobble</li><li>3rd base in codon: multiple different nucleotides encode same AA</li><li>eliminate unnecessary DNA mismatches</li><li>more useful kmer hits</li></ul></li><li>low-complexity regions removed <ul><li>repetitive DNA</li></ul></li><li>exploit kmer <ul><li>generate &#39;neighbors&#39;</li><li>allow mismatches</li><li>ungapped alignment and pick with threshold</li></ul></li><li>identify nearby hits on same diagonal <ul><li>like seed chain align</li><li>extend region to a high-scoring segment pair (HSP)</li></ul></li></ul></li></ul></li></ul><h2 id="_3-1-comparing-sequences" tabindex="-1"><a class="header-anchor" href="#_3-1-comparing-sequences"><span>3.1. Comparing Sequences</span></a></h2><ul><li>multiple sequence alignment <ul><li>identify similarities between sequences</li><li>find conserved patterns (motifs æ¨¡ä½“) in a protein family</li><li>Sum-of-Pairs (SP) score</li><li>dynamic programming</li></ul></li><li>heuristic: progressive alignment <ol><li>construct a matrix</li><li>construct a guide tree</li><li>align sequences</li></ol></li><li>Basic Local Alignment Search Tool (BLAST) <ul><li>one to many query</li></ul><ol><li>split query into words</li><li>find neighbourhood words of similar words</li><li>collect seeds</li><li>extend seeds</li><li>get the best alignment</li></ol></li><li>minhash <ul><li>fingerprint, known as kmers</li><li>big data</li></ul></li><li>Jaccard Coefficient <ul><li>measure the similarity</li><li>proportion</li><li>J(A, B) = |A âˆ© B| / |A âˆª B|</li></ul></li><li>minimizers <ul><li>window</li><li>minimizer: select the minimum kmer from a window according to an order</li></ul></li><li>shazam <ul><li>search audio</li><li>audio, fingerprint, hash, reproducible</li></ul></li><li>limitation <ul><li>window size W and kmer size k trade-off</li><li>longer kmer, more informative</li><li>bigger window, more accurate, less efficient</li></ul></li></ul><h2 id="_3-2-graph-theory" tabindex="-1"><a class="header-anchor" href="#_3-2-graph-theory"><span>3.2. Graph Theory</span></a></h2><ul><li>graph, interaction, biology</li><li>describe and analyse relationships</li><li>nodes, edges</li><li>labels, directional, weights</li><li>adjacent, degree</li><li>walk (edge), path (node), cycle (closed path)</li><li>Eulerian path: visit each edge once and only once</li><li>Hamiltonian path: visit each node exactly once, NP-complete</li><li>tree, usually sparse</li><li>clique, edges between all pairs of nodes</li><li>graph, adjacency list, adjacency matrix</li><li>traversals: depth-first search, breadth-first search</li><li>recursion</li></ul><h2 id="_4-1-network-analysis" tabindex="-1"><a class="header-anchor" href="#_4-1-network-analysis"><span>4.1. Network Analysis</span></a></h2><ul><li>connected?</li><li>distance of nodes?</li><li>density?</li><li>diversity of distribution? (node degree distribution)</li><li>centrality <ul><li>degree</li><li>closeness</li><li>betweenness</li><li>eigenvector</li></ul></li><li>classification of networks <ul><li>scale-free</li><li>small-world</li><li>random</li><li>regular</li></ul></li></ul><h2 id="_4-2-advanced-indexing" tabindex="-1"><a class="header-anchor" href="#_4-2-advanced-indexing"><span>4.2. Advanced Indexing</span></a></h2><ul><li>trie (prefix tree or radix tree) <ul><li>used in word other</li></ul></li><li>suffix tree åç¼€æ ‘ <ul><li>complexity, space</li></ul></li><li>suffix array <ul><li>each item is a word</li></ul></li><li>Burrrows-Wheeler Transform (BWT) <ul><li>popular</li><li>efficient in indexing human genome</li></ul></li></ul><h2 id="_5-evolutionary-trees" tabindex="-1"><a class="header-anchor" href="#_5-evolutionary-trees"><span>5. Evolutionary Trees</span></a></h2><ul><li><p>taxonomy åˆ†ç±»å­¦</p></li><li><p>phylogenetics ç³»ç»Ÿå‘è‚²</p></li><li><p>speciation ç‰©ç§å½¢æˆ</p></li><li><p>multiple sequence alignment</p><ul><li>issue: local not represent global</li></ul></li><li><p>greedy: pair the most similar sequences</p></li><li><p>building a tree</p><ul><li>next merge</li><li>internal node</li><li>branch length</li></ul></li><li><p>UPGMA</p><ul><li>unweighted pair group method with arithmetic mean</li><li><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>O</mi><mo stretchy="false">(</mo><msup><mi>n</mi><mn>3</mn></msup><mo stretchy="false">)</mo></mrow><annotation encoding="application/x-tex">O(n^3)</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1.0641em;vertical-align:-0.25em;"></span><span class="mord mathnormal" style="margin-right:0.02778em;">O</span><span class="mopen">(</span><span class="mord"><span class="mord mathnormal">n</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">3</span></span></span></span></span></span></span></span><span class="mclose">)</span></span></span></span>, n for number of merge, n * n for distance matrix</li><li>distance based</li><li>ultrameric: distance from root to leaf is the same</li></ul></li><li><p>Neighbor joining</p></li><li><p>distance based methods</p><ul><li>pros: simple, flexible, fast, scalable</li><li>cons: sensitive to distance, evolutionary rates not considered, uncertain</li></ul></li><li><p>character based methods</p><ul><li>taxon be described by a number of characters</li></ul></li><li><p>parsimony è§£æ methods</p><ul><li>compute a phylogenetic tree T for a set of sequences that minimizes L(T)</li><li>L(T) is the min of subsitutions requires to explain tree T</li></ul></li><li><p>evaluating trees</p><ul><li>label internal nodes e.g. Hamming distance</li></ul></li><li><p>small parsimony</p><ul><li>given T, find L(T)</li></ul></li><li><p>FItch&#39;s algorithm</p><ul><li>repeatly assign labels</li><li><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>S</mi><mi>v</mi></msub><mo>=</mo><mrow><msub><mi>S</mi><mi>u</mi></msub><mo>âˆª</mo><msub><mi>S</mi><mi>v</mi></msub><mi>i</mi><mi>f</mi><msub><mi>S</mi><mi>u</mi></msub><mo>âˆ©</mo><msub><mi>S</mi><mi>v</mi></msub><mo>=</mo><mi mathvariant="normal">âˆ…</mi><mo separator="true">,</mo><msub><mi>S</mi><mi>u</mi></msub><mo>âˆ©</mo><msub><mi>S</mi><mi>v</mi></msub><mi>o</mi><mi>t</mi><mi>h</mi><mi>e</mi><mi>r</mi><mi>w</mi><mi>i</mi><mi>s</mi><mi>e</mi></mrow></mrow><annotation encoding="application/x-tex">S_v = {S_u \cup S_v if S_u \cap S_v = \emptyset, S_u \cap S_v otherwise}</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8333em;vertical-align:-0.15em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:0.9444em;vertical-align:-0.1944em;"></span><span class="mord"><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">u</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">âˆª</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathnormal">i</span><span class="mord mathnormal" style="margin-right:0.10764em;">f</span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">u</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">âˆ©</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mord">âˆ…</span><span class="mpunct">,</span><span class="mspace" style="margin-right:0.1667em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">u</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">âˆ©</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.03588em;">v</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathnormal">o</span><span class="mord mathnormal">t</span><span class="mord mathnormal">h</span><span class="mord mathnormal" style="margin-right:0.02778em;">er</span><span class="mord mathnormal" style="margin-right:0.02691em;">w</span><span class="mord mathnormal">i</span><span class="mord mathnormal">se</span></span></span></span></span></li></ul></li><li><p>Sankoff&#39;s algorithm</p><ul><li>count smallest number of possible changes needed on a given tree</li><li><span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msub><mi>S</mi><mi>a</mi></msub><mo stretchy="false">(</mo><mi>i</mi><mo stretchy="false">)</mo><mo>=</mo><mi>m</mi><mi>i</mi><mi>n</mi><mo stretchy="false">[</mo><msub><mi>c</mi><mi>i</mi></msub><mi>j</mi><mo>+</mo><msub><mi>S</mi><mi>l</mi></msub><mi>e</mi><mi>f</mi><mi>t</mi><mo stretchy="false">(</mo><mi>j</mi><mo stretchy="false">)</mo><mo stretchy="false">]</mo><mo>+</mo><mi>m</mi><mi>i</mi><mi>n</mi><mo stretchy="false">[</mo><msub><mi>c</mi><mi>i</mi></msub><mi>K</mi><mo>+</mo><msub><mi>S</mi><mi>r</mi></msub><mi>i</mi><mi>g</mi><mi>h</mi><mi>t</mi><mo stretchy="false">(</mo><mi>k</mi><mo stretchy="false">)</mo><mo stretchy="false">]</mo></mrow><annotation encoding="application/x-tex">S_a(i) = min[c_ij + S_left(j)] + min[c_iK + S_right(k)]</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">a</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mopen">(</span><span class="mord mathnormal">i</span><span class="mclose">)</span><span class="mspace" style="margin-right:0.2778em;"></span><span class="mrel">=</span><span class="mspace" style="margin-right:0.2778em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">min</span><span class="mopen">[</span><span class="mord"><span class="mord mathnormal">c</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3117em;"><span style="top:-2.55em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathnormal" style="margin-right:0.05724em;">j</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3361em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.01968em;">l</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathnormal">e</span><span class="mord mathnormal" style="margin-right:0.10764em;">f</span><span class="mord mathnormal">t</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.05724em;">j</span><span class="mclose">)]</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord mathnormal">min</span><span class="mopen">[</span><span class="mord"><span class="mord mathnormal">c</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.3117em;"><span style="top:-2.55em;margin-left:0em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight">i</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathnormal" style="margin-right:0.07153em;">K</span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">+</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:1em;vertical-align:-0.25em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.05764em;">S</span><span class="msupsub"><span class="vlist-t vlist-t2"><span class="vlist-r"><span class="vlist" style="height:0.1514em;"><span style="top:-2.55em;margin-left:-0.0576em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mathnormal mtight" style="margin-right:0.02778em;">r</span></span></span></span><span class="vlist-s">â€‹</span></span><span class="vlist-r"><span class="vlist" style="height:0.15em;"><span></span></span></span></span></span></span><span class="mord mathnormal">i</span><span class="mord mathnormal" style="margin-right:0.03588em;">g</span><span class="mord mathnormal">h</span><span class="mord mathnormal">t</span><span class="mopen">(</span><span class="mord mathnormal" style="margin-right:0.03148em;">k</span><span class="mclose">)]</span></span></span></span></li></ul></li><li><p>build a tree for character based method</p><ul><li>greedy</li><li>sequential addition</li></ul></li><li><p>bootstrapping é‡å¤æŠ½æ ·</p><ul><li>reâ€sample, create new alignment</li></ul></li></ul><h2 id="_6-assembly" tabindex="-1"><a class="header-anchor" href="#_6-assembly"><span>6. Assembly</span></a></h2><ul><li><p>to build contigs è¿ç»­åºåˆ—</p></li><li><p>greedy extension</p><ul><li>for N reads of L length</li><li>O(<span class="katex"><span class="katex-mathml"><math xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><msup><mi>N</mi><mn>3</mn></msup><mo>âˆ—</mo><msup><mi>L</mi><mn>2</mn></msup></mrow><annotation encoding="application/x-tex">N^3 * L^2</annotation></semantics></math></span><span class="katex-html" aria-hidden="true"><span class="base"><span class="strut" style="height:0.8141em;"></span><span class="mord"><span class="mord mathnormal" style="margin-right:0.10903em;">N</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">3</span></span></span></span></span></span></span></span><span class="mspace" style="margin-right:0.2222em;"></span><span class="mbin">âˆ—</span><span class="mspace" style="margin-right:0.2222em;"></span></span><span class="base"><span class="strut" style="height:0.8141em;"></span><span class="mord"><span class="mord mathnormal">L</span><span class="msupsub"><span class="vlist-t"><span class="vlist-r"><span class="vlist" style="height:0.8141em;"><span style="top:-3.063em;margin-right:0.05em;"><span class="pstrut" style="height:2.7em;"></span><span class="sizing reset-size6 size3 mtight"><span class="mord mtight">2</span></span></span></span></span></span></span></span></span></span></span>)</li></ul></li><li><p>overlap layout consensus (OLC)</p><ul><li>Newbler, Celera, Canu</li></ul></li><li><p>sequencing by hybridisation</p><ul><li>microarray</li></ul></li><li><p>De Bruijn Graphs</p><ul><li>kmer</li><li>Eulerian path</li><li>ACGT -&gt; 0,1,2,3</li><li>4^k nodes</li><li>O(kN) space</li><li>O(N) time</li></ul></li><li><p>considerations</p></li><li><p>OLC simplification</p></li><li><p>De Bruijn graph simplification</p></li><li><p>Kmers, coverage, depth</p></li><li><p>assessing assemblies</p></li><li><p>software</p></li></ul><h2 id="_7-dimensionality-reduction" tabindex="-1"><a class="header-anchor" href="#_7-dimensionality-reduction"><span>7. Dimensionality Reduction</span></a></h2><ul><li>approaches <ul><li>feature selection</li><li>feature extraction</li></ul></li><li>principal component analysis (PCA) <ul><li>linear transformation</li><li>eigenvectors</li><li>eigenvalues</li><li>variance</li><li>covariance</li><li>projection</li><li>dimensionality reduction</li></ul></li><li>multidimensional scaling (MDS) <ul><li>distance matrix</li><li>similarity matrix</li><li>stress</li><li>metric MDS</li><li>non-metric MDS</li></ul></li><li>ISOMAP</li><li>t-SNE</li><li>UMAP</li></ul><h2 id="_8-clustering" tabindex="-1"><a class="header-anchor" href="#_8-clustering"><span>8. Clustering</span></a></h2><ul><li>machine learning</li><li>clustering</li><li>distance metrics</li><li>partitional clustering: k-means</li><li>clustering approaches</li><li>density-based clustering</li><li>divisive clustering</li></ul><h2 id="_9-optimisation" tabindex="-1"><a class="header-anchor" href="#_9-optimisation"><span>9. Optimisation</span></a></h2><ul><li>exploring the solution space <ul><li>constraints</li><li>branch and bound</li></ul></li><li>search heuristics <ul><li>gradient descent</li><li>simulated annealing</li><li>genetic algorithms</li></ul></li><li>optimisation</li></ul><h2 id="_10-supervised-machine-learning" tabindex="-1"><a class="header-anchor" href="#_10-supervised-machine-learning"><span>10. Supervised Machine Learning</span></a></h2><ul><li>unsupervised learning <ul><li>associative rule learning</li></ul></li><li>unsupervised vs supervised ML <ul><li>GeoGuessr</li></ul></li><li>supervised ML <ul><li>KNN</li><li>naive Bayes</li><li>SVM</li><li>decision trees</li><li>ensemble methods</li></ul></li><li>genomics example <ul><li>markov models for gene identification</li></ul></li></ul><h2 id="_11-evaluation" tabindex="-1"><a class="header-anchor" href="#_11-evaluation"><span>11. Evaluation</span></a></h2><ul><li>classification</li><li>regression</li></ul><h2 id="guest-lecture-an-introduction-to-machine-learning-and-artificial-intelligence" tabindex="-1"><a class="header-anchor" href="#guest-lecture-an-introduction-to-machine-learning-and-artificial-intelligence"><span>Guest Lecture: An introduction to machine learning and artificial intelligence</span></a></h2></div><!----><!----><!----></div><footer class="vp-page-meta"><!----><div class="vp-meta-item git-info"><div class="update-time"><span class="vp-meta-label">æœ€è¿‘æ›´æ–°: </span><time class="vp-meta-info" datetime="2026-01-07T07:01:42.000Z" data-allow-mismatch>2026/1/7 07:01</time></div><div class="contributors"><span class="vp-meta-label">è´¡çŒ®è€…: </span><!--[--><!--[--><span class="vp-meta-info" title="email: sihong1@student.unimelb.edu.au">shyu216</span><!--]--><!--]--></div></div></footer><nav class="vp-page-nav"><!----><a class="route-link auto-link next" href="/knownoevil/study/mscs/24S2/COMP90018.html" aria-label="COMP90018"><div class="hint">ä¸‹ä¸€é¡µ<span class="arrow end"></span></div><div class="link">COMP90018<i class="vp-icon fas fa-book-open" sizing="height"></i></div></a></nav><!----><!----><!--]--></main><!--]--><footer class="vp-footer-wrapper" vp-footer><div class="vp-footer">A place to jot down the picked up.</div><div class="vp-copyright">Copyright Â© 2026 Dale </div></footer></div><!--]--><!--]--><!--[--><!----><!----><!--[--><!--]--><!--]--><!--]--></div>
    <script type="module" src="/knownoevil/assets/app-Dk-GzZ3g.js" defer></script>
  </body>
</html>
