---
title: COMP90014 Algorithms for Bioinformatics
shortTitle: COMP90014
order: 2
icon: book-open
category:
  - UniMelb
  - 24S2
tag:
  - Algorithm
  - Bioinformatics
---

## 1. Indexing

- common classes of algorithms
  - brute force
  - divide and conquer
  - recursive
  - dynamic programming
  - greedy
- data in bioinformatics
  - FASTA, sequence, genome
  - FASTQ, with quality, read
  - BAM/SAM, alignment
  - Matrix
  - VCF, genetic variants
- data type
  - primitives: int, float, char
  - collections: array, list, set, map
  - derived types: class, enum, tuple
- data structure
  - arrangement of data in algorithm
  - queue, stack, heap, tree, graph
- evaluation
  - correctness
  - efficiency
  - asymptotic complexity 渐近复杂度 (big O notation)
  - tractability 可解性 (P, NP, NP-complete)
- when complex
  - heuristic
  - approximation
- indexing
  - pattern matching
  - see if something exists
  - linear search O(n)
  - binary search O(log n)
  - hash table O(1)
- double hashing (perfect hashing): use another hash function to treat collision
- kmer
  - breaking sequence into smaller pieces
  - key of genomic data in indexing

## 2. Sequence Alignment and Mapping

- why align sequence
  - compare sequence
  - assess similarity
  - find evolutionary relationships
  - identify functionally conserved sequences
  - homology 同源性，指不同生物体的基因或蛋白质序列由于共同的祖先而具有相似性，有助于理解生物进化、基因功能和结构
    - homologs 同源基因
    - orthologs 直系 separated by speciation
    - paralogs 旁系 separated by duplication
  - phylogeny 系统发育 infer the evolutionary relationships among sequences
  - protein function, find regions of known function, infer function
  - conservation, find conserved regions 在不同物种或不同基因组中，某些基因或蛋白质序列在进化过程中保持不变或变化很小的区域，通常这些序列在生物学功能上具有重要性
  - data searching
  - de novo assembly, reconstruct the original sequence
  - read mapping, infer variants
- how to align
  - pairwise
  - BLAST
- Hamming distance
  - for equal length sequences, count the number of different characters
- kmer distance
  - for unequal length sequences, count the number of different kmers
- Levenshtein distance (pairwise)
  - edit distance
  - 最小编辑距离，指两个字符串之间，由一个转换成另一个所需的最少编辑操作次数
  - count mismatch, insertion, deletion as transformations
  - argmin of 
    - (i-1, j-1) + (missmatch 1 or match 0)
    - (i-1, j) + (gap 1)
    - (i, j-1) + (gap 1)
  - return the bottom-right cell of the matrix
- pairwise distance with penalties
  - apply penalties
    - i.e. match 0, mismatch -1, gap -2
  - return the bottom-right cell of the matrix
- Needleman-Wunsch algorithm (global)
  - return the bottom-right cell of the matrix
- Smith-Waterman algorithm (local)
  - return the maximum value in the matrix
- semi-global alignment
  - clipped by best offset
  - Needleman-Wunsch variant
    - start from max cell, end at reaching top row/left column
- (scoring) substitution matrix 替换矩阵
  - assign scores to each pair of characters
  - why not use fixed scores
    - two types of DNA mutations 两种DNA突变
      - transition, purine to purine or pyrimidine to pyrimidine, frequently (A↔G, C↔T)
      - transversion, purine to pyrimidine or vice versa, less likely (A↔C, A↔T, G↔C, G↔T)
  - it reflect
    - chemical similarity
    - observed mutation frequencies
    - how proteins sequence evolve
- Point Accepted Mutation (PAM) matrix
  - the number represents the evolutionary distance between the sequences
  - higher numbers denote greater distances
  - limitation
    - only for closely related sequences and small dataset
    - not consider different rates of evolution
- BLOcks SUbstitution Matrix (BLOSUM) matrix
  - scores derived from frequencies of substitutions in blocks of ungapped local alignments
  - BLOSUM62: 62% identity
- gap panalty
  - negative score
  - fixed / affine
- read mapping
  - seed extension
    - combine indexing and alignment
    - process
      1. index kmer
      2. find seed
      3. extend seed using alignment
      4. return best
  - seed chain align
    - for long, noisy sequences
    - process
      1. find seeds
      2. identify colinear chains
      3. base-level alignments to fill gaps
      4. return best
- Basic Local Alignment Search Tool (BLAST)
  - efficient heuristic
    - DNA translated to protein seq for use
      - Amino Acid (AA) sequences more conserved than DNA
      - degeneracy / codon wobble
      - 3rd base in codon: multiple different nucleotides encode same AA
      - eliminate unnecessary DNA mismatches
      - more useful kmer hits
    - low-complexity regions removed
      - repetitive DNA
    - exploit kmer
      - generate 'neighbors'
      - allow mismatches
      - ungapped alignment and pick with threshold
    - identify nearby hits on same diagonal
      - like seed chain align
      - extend region to a high-scoring segment pair (HSP)

## 3.1. Comparing Sequences

- multiple sequence alignment
  - identify similarities between sequences
  - find conserved patterns (motifs 模体) in a protein family
  - Sum-of-Pairs (SP) score
  - dynamic programming
- heuristic: progressive alignment
  1. construct a matrix
  2. construct a guide tree
  3. align sequences
- Basic Local Alignment Search Tool (BLAST)
  - one to many query
  1. split query into words
  2. find neighbourhood words of similar words
  3. collect seeds
  4. extend seeds
  5. get the best alignment
- minhash
  - fingerprint, known as kmers
  - big data
- Jaccard Coefficient
  - measure the similarity
  - proportion
  - J(A, B) = |A ∩ B| / |A ∪ B|
- minimizers
  - window
  - minimizer: select the minimum kmer from a window according to an order
- shazam
  - search audio
  - audio, fingerprint, hash, reproducible
- limitation
  - window size W and kmer size k trade-off
  - longer kmer, more informative
  - bigger window, more accurate, less efficient

## 3.2. Graph Theory

- graph, interaction, biology
- describe and analyse relationships
- nodes, edges
- labels, directional, weights
- adjacent, degree
- walk (edge), path (node), cycle (closed path)
- Eulerian path: visit each edge once and only once
- Hamiltonian path: visit each node exactly once, NP-complete
- tree, usually sparse
- clique, edges between all pairs of nodes
- graph, adjacency list, adjacency matrix
- traversals: depth-first search, breadth-first search
- recursion

## 4.1. Network Analysis

- connected?
- distance of nodes?
- density?
- diversity of distribution? (node degree distribution)
- centrality
  - degree
  - closeness
  - betweenness
  - eigenvector
- classification of networks
  - scale-free
  - small-world
  - random
  - regular

## 4.2. Advanced Indexing

- trie (prefix tree or radix tree)
  - used in word retrieval
- suffix tree
  - complexity, space
- suffix array

## 5. Evolutionary Trees

- taxonomy 分类学
- phylogenetics 系统发育
- speciation 物种形成
- multiple sequence alignment
  - issue: local not represent global
- greedy: pair the most similar sequences
- building a tree
  - next merge
  - internal node
  - branch length
- UPGMA
  - unweighted pair group method with arithmetic mean
  - $O(n^3)$, n for number of merge, n * n  for distance matrix